include/net/dsa.h

```diff
@@ -0,0 +1,222 @@
+ *
+ * Returns:
+ *  - NET_DROP, if packet was invalid, rejected or we want the stack to free it.
+ *    In this case the core stack will free the packet.
+ *  - NET_OK, if the packet was accepted, in this case the ownership of the
+ *    net_pkt goes to callback and core network stack will forget it.
+ */
+typedef enum net_verdict (*dsa_net_recv_cb_t)(struct net_if *iface,
+					      struct net_pkt *pkt);
+
+/**
+ * @brief Register DSA Rx callback functions
+ *
+ * @param iface Network interface
+ * @param cb Receive callback function
+ *
+ * @return 0 if ok, < 0 if error
+ */
+int dsa_register_recv_callback(struct net_if *iface, dsa_net_recv_cb_t cb);
+
+/**
+ * @brief Parse DSA packet
+ *
+ * @param iface Network interface (master)
+ * @param pkt Network packet
+ *
+ * @return Return 1 if packet needs to be altered in any way
+ */
+int net_dsa_recv(struct net_if *iface, struct net_pkt *pkt);
+
+/**
+ * @brief Set DSA interface to packet
+ *
+ * @param iface Network interface (master)
+ * @param pkt Network packet
+ *
+ * @return Return the slave network interface
+ */
+struct net_if *dsa_recv_set_iface(struct net_if *iface, struct net_pkt **pkt);
+
+/**
+ * @brief DSA helper function to get context
+ *
+ * This is the helper function to provide DSA context pointer.
+ * (master and LAN interfaces shall have pointer to struct dsa_context
+ * available)
+ *
+ * @param iface Network interface (master)
+ *
+ * Returns:
+ *  - 0 if ok (packet sent via master iface), < 0 if error
+ */
+struct dsa_context *dsa_get_context(struct net_if *iface);
+
+/**
+ * @brief Pointer to master interface send function
+ */
+typedef int (*dsa_send_t)(const struct device *dev, struct net_pkt *pkt);
+
+/**
+ * @brief DSA helper function to register transmit function for master
+ *
+ * @param iface Network interface (master)
+ * @param fn Pointer to master interface send method
+ *
+ * Returns:
+ *  - 0 if ok, < 0 if error
+ */
+int dsa_register_master_tx(struct net_if *iface, dsa_send_t fn);
+
+/**
+ * @brief DSA helper function to check if port is master
+ *
+ * @param iface Network interface (master)
+ *
+ * Returns:
+ *  - true if ok, false otherwise
+ */
+bool dsa_is_port_master(struct net_if *iface);
+
+/** DSA context data */
+struct dsa_context {
+	/** Pointers to all DSA slave network interfaces */
+	struct net_if *iface_slave[NET_DSA_PORT_MAX_COUNT];
+
+	/** Pointer to DSA master network interface */
+	struct net_if *iface_master;
+
+	/** DSA related work (e.g. monitor if network interface is up) */
+	struct k_delayed_work dsa_work;
+
+	/** The switch_id, which equals to the reg property number from
+	 * DTS is used to distinct between many connected switches.
+	 */
+	uint8_t switch_id;
+
+	/** Number of slave ports in the DSA switch */
+	uint8_t num_slave_ports;
+
+	/** Status of each port */
+	bool link_up[NET_DSA_PORT_MAX_COUNT];
+
+	/** DSA specific API callbacks - filled in the switch IC driver */
+	struct dsa_api *dapi;
```

Comment:
Minor nit, but could you place this pointer after the `iface_master` one, there is less chance of leaving holes in the struct that way.